File: Backend-RS/models/mensaje.py
from sqlmodel import Field, SQLModel, Relationship
from typing import Optional, List
import datetime
from sqlalchemy import Column, Text, DateTime

class MensajeBase(SQLModel):
    contenido: str
    Leido: bool = False

class Mensaje(MensajeBase, table=True):
    IdMensaje: Optional[int] = Field(default=None, primary_key=True)
    IdConversacion: int = Field(foreign_key="conversacion.IdConversacion")
    IdUsuario: int = Field(foreign_key="usuario.IdUsuario") 
    FechaEnvio: datetime.datetime = Field(
        default_factory=datetime.datetime.utcnow,
        sa_column=Column(DateTime(timezone=True))
    )
    conversacion: "Conversacion" = Relationship(back_populates="mensajes")
    usuario: "Usuario" = Relationship(back_populates="mensajes")

class MensajeCreate(MensajeBase):
    IdConversacion: int
    IdUsuario: int

class MensajePublic(MensajeBase):
    IdMensaje: int
    IdConversacion: int
    IdUsuario: int
    FechaEnvio: datetime.datetime
    usuario: Optional["UsuarioPublic"] = None
    conversacion:Optional["ConversacionPublic"] = None


class MensajeUpdate(SQLModel):
    contenido: Optional[str] = None
    Leido: Optional[bool] = None



from models.conversacion import ConversacionPublic
from models.usuario import UsuarioPublic
MensajePublic.model_rebuild()        

File: frontendrs/lib/data/models/mensaje.dart
class Mensaje {
  final int idMensaje;
  final String contenido;
  final bool leido;
  final int idConversacion;
  final int idUsuario;
  final DateTime fechaEnvio;

  const Mensaje({
    required this.idMensaje,
    required this.contenido,
    required this.leido,
    required this.idConversacion,
    required this.idUsuario,
    required this.fechaEnvio,
  });

  factory Mensaje.fromJson(Map<String, dynamic> json) {
    return Mensaje(
      idMensaje: json['idMensaje'] ?? 0,
      contenido: json['contenido'] ?? "",
      leido: json['leido'] ?? "",
      idConversacion: json['idConversacion'] ?? 0,
      idUsuario: json['idUsuario'] ?? 0,
      fechaEnvio: DateTime.parse(json['fechaEnvio']),
    );
  }

  Map<String, dynamic> toJson() => {
    'idMensaje': idMensaje,
    'contenido': contenido,
    'leido': leido,
    'idConversacion': idConversacion,
    'idUsuario': idUsuario,
    'fechaEnvio': fechaEnvio.toIso8601String(),
  };
}
